notifier/notify_windows.go

```diff
@@ -2,351 +2,342 @@ package main
-		if elem == "" {
-			continue
+		// Set Portmaster icon.
+		iconLocation, err := ensureAppIcon()
+		if err != nil {
+			log.Warningf("notify: failed to write icon: %s", err)
 		}
-
-		elemSplit := strings.SplitN(elem, "=", 2)
-		if len(elemSplit) != 2 {
-			log.Warningf("failing to parse snoretoast-Response %s into Key=Value-Pair; Response: %s", elem, in)
-			continue
+		iconPathUTF, _ := windows.UTF16PtrFromString(iconLocation)
+		iconPathP := unsafe.Pointer(iconPathUTF)
+		_, _, _ = lib.setImage.Call(notificationTemplate, uintptr(iconPathP))
+
+		// Set all the required actions
+		for _, action := range n.AvailableActions {
+			textUTF, _ := windows.UTF16PtrFromString(action.Text)
+			textP := unsafe.Pointer(textUTF)
+			_, _, _ = lib.addButton.Call(notificationTemplate, uintptr(textP))
 		}
 
-		switch elemSplit[0] {
-		case "action":
-			ret.action = elemSplit[1]
-		case "button":
-			ret.button = elemSplit[1]
-		case "notificationId":
-			ret.id = elemSplit[1]
-		case "version", "pipe": // not needed: do nothing
-		default:
-			log.Infof("failed to parse key %s from SnoreToast-Response into struct: receivedkey is unknown to Portmaster (%s)", elemSplit[0], elem)
+		// Show notification and delete c notification object
+		rc, _, err := lib.showNotification.Call(notificationTemplate)
+		if int64(rc) == -1 {
+			log.Errorf("notify: failed to show notification: %q", err)
 		}
-	}
+		n.systemID = NotificationID(rc)
 
-	if ret.action == "" {
-		log.Errorf("missing attribute action in SnoreToast Response: %s", in)
+		// Link system id to the notification object
+		notifsByIDLock.Lock()
+		notifsByID[NotificationID(rc)] = n
+		notifsByIDLock.Unlock()
 	}
+}
 
-	if ret.id == "" {
-		log.Errorf("missing attribute id in SnoreToast Response: %s", in)
+// Cancel cancels the notification.
+func (n *Notification) Cancel() {
+	if lib == nil {
+		log.Errorf("notify: notification library not properly loaded")
+		return
 	}
 
-	return ret
-}
+	if !libInitialized {
+		log.Error("notify: not initialized")
+	}
 
-func (n *Notification) buildSnoreToastButtonArgument() string {
-	n.Lock()
-	defer n.Unlock()
+	// Lock notification
+	n.RLock()
+	defer n.RUnlock()
 
-	temp := make([]string, 0, len(n.AvailableActions))
-	for _, action := range n.AvailableActions {
-		if !action.IsSupported() {
-			continue
-		}
+	// Lock library
+	lib.Lock()
+	defer lib.Unlock()
 
-		if err := verifySnoreToastArgumentSyntax(action.Text); err != nil {
-			log.Errorf("failed to build SnoreToast Button-Argument: failed to validate Text for %+v: %s", action, err)
-			continue
-		}
+	// the hide function deletes the notification
+	_, _, _ = lib.hideNotification.Call(uintptr(n.systemID))
 
-		if action.Text != "" {
-			temp = append(temp, action.Text)
-		}
-	}
-	return strings.Join(temp, ";")
+	notifsByIDLock.Lock()
+	delete(notifsByID, n.systemID)
+	notifsByIDLock.Unlock()
 }
 
-func runSnoreToastCmd(cmd *exec.Cmd) {
-	exit, err := execCmd(cmd)
+func actionListener() {
```

Comment:
Not sure if this function name is still fitting. Was confusing to me. Maybe document or change.